netty/src/main/java/io/grpc/netty/NettyAdaptiveCumulator.java

```diff
@@ -155,25 +163,20 @@ static void mergeWithCompositeTail(
         // Take ownership of the tail.
         newTail = tail.retain();
 
-        // TODO(https://github.com/netty/netty/issues/12844): remove when we use Netty with
-        //   the issue fixed.
-        // In certain cases, removing the CompositeByteBuf component, and then adding it back
-        // isn't idempotent. An example is provided in https://github.com/netty/netty/issues/12844.
-        // This happens because the buffer returned by composite.component() has out-of-sync
-        // indexes. Under the hood the CompositeByteBuf returns a duplicate() of the underlying
-        // buffer, but doesn't set the indexes.
-        //
-        // To get the right indexes we use the fact that composite.internalComponent() returns
-        // the slice() into the readable portion of the underlying buffer.
-        // We use this implementation detail (internalComponent() returning a *SlicedByteBuf),
-        // and combine it with the fact that SlicedByteBuf duplicates have their indexes
-        // adjusted so they correspond to the to the readable portion of the slice.
-        //
-        // Hence composite.internalComponent().duplicate() returns a buffer with the
-        // indexes that should've been on the composite.component() in the first place.
-        // Until the issue is fixed, we manually adjust the indexes of the removed component.
-        ByteBuf sliceDuplicate = composite.internalComponent(tailComponentIndex).duplicate();
-        newTail.setIndex(sliceDuplicate.readerIndex(), sliceDuplicate.writerIndex());
+        // In older versions some buffers don't support arrayOffset(), so for performance
+        // we don't want to rely on the arrayOffset when we know the alternate path works
+        if (usingPre4_1_111_Netty) {
+          // Because of the way duplicate() works in older netty versions this works.  It stops
+          // working in 4.1.111 because they changed what duplicate returns.
```

Comment:
I don't think it's fair to put blame on changes in Netty. It seems that this might have always been violating the API contract of Netty's CompositeByteBuf. The [javadoc of CompositeByteBuf.internalComponent](https://github.com/netty/netty/blob/8b803abf36a757911e77f747b338b55a02822022/buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java#L1570-L1571) says that "updating the indexes of the returned buffer will lead to an undefined behavior of this buffer". That's exactly what happened here.